newrelic/hooks/mlmodel_langchain.py

```diff
@@ -255,6 +259,476 @@ def wrap_similarity_search(wrapped, instance, args, kwargs):
+            full_chat_completion_summary_dict = {"metadata.%s" % key: value for key, value in metadata.items()}
+            full_chat_completion_summary_dict.update(
+                {
+                    "id": completion_id,
+                    "appName": app_name,
+                    "conversation_id": conversation_id,
+                    "span_id": span_id,
+                    "trace_id": trace_id,
+                    "transaction_id": transaction.guid,
+                    "vendor": "langchain",
+                    "ingest_source": "Python",
+                    "virtual_llm": True,
+                    "request_id": run_id,
+                    "duration": ft.duration,
+                    "response.number_of_messages": len(input_message_list),
+                    "tags": tags,
+                    "error": True,
+                }
+            )
+
+            transaction.record_custom_event("LlmChatCompletionSummary", full_chat_completion_summary_dict)
+
+            create_chat_completion_message_event(
+                transaction,
+                settings.app_name,
+                input_message_list,
+                completion_id,
+                span_id,
+                trace_id,
+                run_id,
+                conversation_id,
+                [],
+                message_ids,
+            )
+
+            raise
+
+    if not return_val:
+        return return_val
+
+    response = return_val
+    run_manager_info = getattr(transaction, "_nr_run_manager_info", {})
+    if hasattr(transaction, "_nr_run_manager_info"):
+        del transaction._nr_run_manager_info
+    run_id = run_manager_info.get("run_id", "")
+    metadata = run_manager_info.get("metadata", {})
+    tags = run_manager_info.get("tags", "")
+
+    input_message_list = [_input]
+    output_message_list = []
+    response_len = 1
+    try:
+        output_message_list = [response[0]] if response else []
+    except:
+        try:
+            output_message_list = [str(response)]
+        except:
+            _logger.warning("Unable to capture response. No response message event will be captured.")
+
+    # Make sure the builtin attributes take precedence over metadata attributes.
+    full_chat_completion_summary_dict = {"metadata.%s" % key: value for key, value in metadata.items()}
+    full_chat_completion_summary_dict.update(
+        {
+            "id": completion_id,
+            "appName": app_name,
+            "conversation_id": conversation_id,
+            "span_id": span_id,
+            "trace_id": trace_id,
+            "transaction_id": transaction.guid,
+            "vendor": "langchain",
+            "ingest_source": "Python",
+            "virtual_llm": True,
+            "request_id": run_id,
+            "duration": ft.duration,
+            "response.number_of_messages": len(input_message_list) + len(output_message_list),
+            "tags": tags or "",
```

Comment:
Yes because tags is defaulted to None in the run function's signature so if the customer doesn't pass in tags, tags will be None not "". I suppose we could handle this case up above though instead or here.
```
    def run(
        self,
        *args: Any,
        callbacks: Callbacks = None,
        tags: Optional[List[str]] = None,
        metadata: Optional[Dict[str, Any]] = None,
        **kwargs: Any,
    ) -> Any:
```